ember-primitives/src/components/one-time-password.gts

```diff
@@ -0,0 +1,181 @@
+
+  for (let [key, value] of formData.entries()) {
+    if (key.startsWith('code')) {
+      code += value;
+    }
+  }
+
+  submit({
+    code
+  });
+};
+
+const reset = (event: Event) => {
+  assert('[BUG]: reset called without an event.target', event.target instanceof HTMLElement);
+
+  let form = event.target.closest('form');
+
+  assert('Form is missing. Cannot use <Reset> without being contained within a <form>', form instanceof HTMLFormElement);
+
+  form.reset();
+}
+
+const autoAdvance = (event: Event) => {
+  assert('[BUG]: reset called without an event.target', event.target instanceof HTMLInputElement);
+
+  let value = event.target.value;
+
+  if (value.length === 1 && /\d/.test(value)) {
+    let nextElement = event.target.nextElementSibling;
+
+    if (nextElement instanceof HTMLElement) {
+      nextElement.focus?.();
+    }
+
+    return;
+  }
+
+  const digits = value;
+  let i = 0;
+  let currElement: HTMLInputElement | null = event.target;
+
+  while(currElement) {
+    currElement.value = digits[i++] || '';
+
+    if (currElement.nextElementSibling instanceof HTMLElement) {
+      currElement = currElement.nextElementSibling;
+    }
+  }
+}
+
+function fieldsFor(length: number) {
+  return new Array(length);
+}
+
+function labelFor(inputIndex: number, labelFn: undefined | ((index: number) => string)) {
+  if (labelFn) {
+    return labelFn(inputIndex);
+  }
+
+  return `Please enter OTP character ${inputIndex + 1}`;
+}
+
+function defaultOf<T>(input: T | undefined, fallback: T) {
+  return input ?? fallback;
+}
+
+const Input: TOC<{
+  /**
+    * Any attributes passed to this component will be applied to each input.
+    */
+  Element: HTMLInputElement;
+  Args: {
+    length: number;
+    /**
+      * To Customize the label of the input fields, you may pass a function.
+      * By default, this is `Please enter OTP character ${index + 1}`.
+      */
+    labelFn?: (index: number) => string;
+  }
+}> = <template>
+    {{#each (fieldsFor (defaultOf @length DEFAULT_LENGTH)) as |_field i|}}
+      <input
+        name="code{{i}}"
+        type="number"
+        max="9" min="0" step="1"
+        ...attributes
+        aria-label={{labelFor i @labelFn}}
+        {{on 'input' autoAdvance}}
+      >
+    {{/each}}
+</template>;
+
+const Submit: TOC<{
+  Element: HTMLButtonElement;
+  Blocks: { default: [] }
+}> = <template>
+    <button type="submit" ...attributes>Submit</button>
+</template>;
+
+const Reset: TOC<{
+  Element: HTMLButtonElement;
+  Blocks: { default: []}
+}> = <template>
+    <button type="button" {{on 'click' reset}} ...attributes>{{yield}}</button>
+</template>;
+
+export const OTPInput: TOC<{
+  /**
+    * The overall OTP Input is in its own form.
+    * Modern UI/UX Patterns usually have this sort of field
+  * as its own page, thus within its own form.
+    *
+    * By default, only the 'submit' event is bound, and is
+  * what calls the `@onSubmit` argument.
+    */
+  Element: HTMLFormElement,
```

Comment:
I could imagine use case where the OTP is part of a bigger form:

```
Login: [ enter email ]
Password: [ enter password ]
Token: [ enter token from security device ]

[Cancel]  [ Login ]
```

Could the `OTPInput` be composed in such a form?

As a component user, I would expect to be able to do something like this:
```
<form ....submit handler...>
   ... email and password fields ...
   <label>Security Token:
   <OTPInput @value={{this.otpValue}} @onchange={{this.updateOTPValue}} /></label>
   
   ... cancel and reset buttons ...
</form>
```

In other words, the `OTPInput` would not own the form element.

Reply:
maybe this is the difference between OTP and OTPInput?

OTPInput can be "use anywhere, even in an existing form".
And OTP can be the standalone pre-wired thing?

Reply:
This is now implemented!